icu4c/source/i18n/tibetancal.cpp

```diff
@@ -0,0 +1,538 @@
+// © 2016 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+/*
+ * Copyright (C) 2003-2021, International Business Machines Corporation
+ * and others. All Rights Reserved.
+ ******************************************************************************
+ *
+ * File TIBETANCAL.CPP
+ *****************************************************************************
+ */
+
+#include "tibetancal.h"
+#include <stdlib.h>
+#if !UCONFIG_NO_FORMATTING
+
+#include "umutex.h"
+#include <float.h>
+#include "gregoimp.h" // Math
+#include "uhash.h"
+
+// Debugging
+#ifdef U_DEBUG_TIBETANCAL
+#include <stdio.h>
+#include <stdarg.h>
+
+#endif
+
+U_NAMESPACE_BEGIN
+
+// Implementation of the TibetanCalendar class
+
+//-------------------------------------------------------------------------
+// Constructors...
+//-------------------------------------------------------------------------
+
+
+TibetanCalendar* TibetanCalendar::clone() const {
+  return new TibetanCalendar(*this);
+}
+
+TibetanCalendar::TibetanCalendar(const Locale& aLocale, UErrorCode& success)
+  :   Calendar(TimeZone::forLocaleOrDefault(aLocale), aLocale, success)
+{
+  setTimeInMillis(getNow(), success); // Call this again now that the vtable is set up properly.
+}
+
+TibetanCalendar::TibetanCalendar(const TibetanCalendar& other) : Calendar(other) {
+}
+
+TibetanCalendar::~TibetanCalendar()
+{
+}
+
+/**
+ * return the calendar type, "tibetan" or "tibetan-tsurphu".
+ *
+ * @return calendar type
+ * @internal
+ */
+const char *TibetanCalendar::getType() const { 
+   const char *sType = NULL;
+
+   switch (cType) {
+    case PHUGPA:
+      sType = 'tibetan';
+      break;
+    case TSURPHU:
+      sType = 'tibetan-tsurphu';
+      break;
+    default:
+      UPRV_UNREACHABLE; // out of range
+   }
+   return sType;
+}
+
+
+/**
+ * Sets Tibetan calendar calculation type used by this instance.
+ *
+ * @param type    The calendar calculation type, <code>PHUGPA</code> to use the phugpa
+ *                calendar, <code>TSURPHU</code> to use the tsurphu calendar.
+ * @internal
+ */
+void TibetanCalendar::setCalculationType(ECalculationType type, UErrorCode &status)
+{
+    if (cType != type) {
+        // The fields of the calendar will become invalid, because the calendar
+        // rules are different
+        UDate m = getTimeInMillis(status);
+        cType = type;
+        clear();
+        setTimeInMillis(m, status);
+    }
+}
+
+
+/**
+* Returns <code>true</code> if this object is using the phugpa
+* calendar, or <code>false</code> if using the tsurphu calendar.
+*/
+UBool TibetanCalendar::isPhugpa() {
+    return (cType == PHUGPA);
+}
+
+
+
+//-------------------------------------------------------------------------
+// Assorted calculation utilities
+//-------------------------------------------------------------------------
+
+// fixed tables
+int32_t SUN_TAB[] = {0, 6, 10, 11};
```

Comment:
Please use constexpr and add comments ref point to  [Janson]
For example, change it to
```
constexpr int32_t SUN_TAB[] = {0, 6, 10, 11}; // as in [Janson] (7.21)
constexpr int32_t MOON_TAB[] = {0, 5, 10, 15, 19, 22, 24, 25}; // as in [Janson] (7.18)
```

Reply:
I understand in the paper they try to use a smaller table, but without the folding the size is 12 and 28 which is not very big anywa. Why don't we just use 12 and 28 items so later on we do not need all those if check to perform better.
```

int32_t moonTab(int32_t i) {
    // as in [Janson] (7.18)
    static constexpr int32_t MOON_TAB[] = {
        0,  5,  10,  15,  19,  22,  24,  25,  24,  22,  19,  15,  10,  5,  
        0, -5, -10, -15, -19, -22, -24, -25, -24, -22, -19, -15, -10, -5 
    }; 
    return MOON_TAB[(i % 28 + 28) % 28];
}
int32_t sunTab(int32_t i) {
    // as in [Janson] (7.21)
    constexpr int32_t SUN_TAB[] = {
        0, 6, 10, 11, 10, 6, 0, -6, -10, -11, -10, - 6
    }; 
    return SUN_TAB[(i % 12 + 12) % 12];
}
```